sparkly/testing.py

```diff
@@ -120,28 +147,122 @@ def tearDown(self):
         """
         if fields:
             actual_df = actual_df.select(*fields)
 
         actual_rows = actual_df.collect() if hasattr(actual_df, 'collect') else actual_df
         actual_data = [row.asDict(recursive=True) for row in actual_rows]
 
-        if ordered:
-            self.assertEqual(actual_data, expected_data)
-        else:
-            try:
-                self.assertCountEqual(actual_data, expected_data)
-            except AttributeError:
-                self.assertItemsEqual(actual_data, expected_data)
+        return self.assertEqual(
+            actual_data,
+            expected_data,
+            ignore_order=not ordered,
+            max_recursion_depth=1,
+            places=places,
+        )
+
+    def assertEqual(self,
+                    first,
+                    second,
+                    msg=None,
+                    ignore_order=False,
+                    max_recursion_depth=None,
+                    places=None):
+        """Assert equal on steroids.
+
+        Extend this classic function signature to work better with
+        comparisons involving dataframes, rows, dictionaries and lists by
+        ignoring the order of mutable nested structures recursively and
+        rounding up floats when necessary.
+
+        Args:
+            first, second, msg: see unittest.TestCase.assertCountEqual
+            ignore_order (bool|False): ignore the order in all structures
+                that are inherently oredered (e.g., lists, mock calls).
+            max_recursion_depth (int|None): if ignore_order is true, do
+                ignore order up to this nesting level (exclusive). Setting
+                this to 0 is same to setting ignore_order=False, 1 means
+                ignore order only at the top level and so on. Default is
+                ignore order arbitrarily deep.
+            places (int|None): Compare floats up to given decimal places.
+
+        Returns:
+            None, iff the two objects are equal.
+
+        Raises
+            AssertionError, iff the two objects are not equal. See
+            unittest.TestCase.assertEqual for details.
+        """
+        def _create_key(entry):
+            if isinstance(entry, type(mock.call)):
+                _, call_args, call_kwargs = entry
+                entry = list(call_args) + list(call_kwargs.items())
+            if isinstance(entry, collections.Iterable) and not isinstance(entry, six.string_types):
+                return entry is None, tuple(_create_key(x) for x in entry)
+            return entry is None, entry
+
+        def _nicefy(obj, recursion_depth=0):
+            if isinstance(obj, float):
+                if math.isnan(obj):
+                    # assertEqual does not recognize two nans as being
+                    # equal otherwise
+                    obj = 'nan'
+                elif places is not None:
+                    obj = round(obj, places)
```

Comment:
While I see how rounding makes this whole thing easier, I don't think it's the best we can do with fuzzy float comparison.  In general it's just not a good way to do tolerance checking, especially if your values can differ by large orders of magnitude.  A percentage-based approach would be far more resilient: "values must be within 0.001%" is a much better way to specify approximate matching because this handles differing scale in a graceful way.  I admit that most of our use cases to date have been simple calculations that we can control, but 1) this is a general-purpose tool and 2) we will probably start introducing more complex calculations which will need tolerance-based comparison rather than this simpler approach.